### PR TITLE
[xtask] allow overriding `clippy` message format

### PR DESCRIPTION
**Pull request by hawkw**
Currently, I've configured VS Code to run checks on save using the same
list of lints checked by `cargo xtask clippy` with a
`.vscode/settings.json` like this:

```json
{
    "rust-analyzer.check.overrideCommand": [
        "cargo",
        "clippy",
        "--all-targets",
        "--message-format=json",
        "--",
        "--allow",
        "clippy::style",
        "--warn",
        "clippy::disallowed_macros",
        "--warn",
        "clippy::disallowed_methods",
        "--warn",
        "clippy::disallowed_names",
        "--warn",
        "clippy::disallowed_script_idents",
        "--warn",
        "clippy::disallowed_types",
        "--warn",
        "clippy::iter_cloned_collect",
        "--warn",
        "clippy::iter_next_slice",
        "--warn",
        "clippy::iter_nth",
        "--warn",
        "clippy::iter_nth_zero",
        "--warn",
        "clippy::iter_skip_next",
        "--warn",
        "clippy::len_zero",
        "--warn",
        "clippy::redundant_field_names",
        "--warn",
        "clippy::cast_lossless"
    ]
}
```

Having to reproduce the list of lints in the `settings.json` is a bit
unfortunate, though, because if a developer wishes to keep their list of
enabled lints in sync with those run by `cargo xtask clippy`, they must
manually update the VS Code config when new lints are added. It would be
nicer if the VS Code config could just run `cargo xtask clippy` on save.

However, this isn't currently possible, as rust-analyzer relies on the
JSON output format for `cargo check`/`cargo clippy`, so it needs to pass
the `--message-format=json` flag to `cargo clippy`. This flag is not
currently propagated by `cargo xtask clippy`.

Therefore, I've updated the xtask to accept a `--message-format`
argument that can be passed through to the underlying `cargo clippy`
command. Now, the VS Code config can be changed to just:

```json
{
    "rust-analyzer.check.overrideCommand": [
        "cargo",
        "xtask",
        "clippy",
        "--message-format=json",
    ]
}
```

### CODE REVIEW

**Comment by hawkw** [comment]
I'm also open to checking in a `.vscode/config.json` that configures VS Code to run `cargo xtask clippy` on save, but IIRC, someone mentioned that there are some issues with committing VS Code config files? Alternatively, we could add this to developer documentation someplace.

**Review by smklein — APPROVED** [review]

**Comment by iliana** [comment]
Disabled auto merge for merge freeze

**Comment by hawkw** [comment]
> Disabled auto merge for merge freeze

ah, whoops, sorry!

File: dev-tools/xtask/src/clippy.rs [hunk 1]
```diff
@@ -13,6 +13,9 @@ pub struct ClippyArgs {
     /// Automatically apply lint suggestions.
     #[clap(long)]
     fix: bool,
+    /// Error format passed to `cargo clippy`.
+    #[clap(long)]
+    message_format: Option<String>,
```
Comment (sunshowers) [inline dev-tools/xtask/src/clippy.rs:18]: Maybe `value_name = "FMT"` to match what cargo produces?

File: dev-tools/xtask/src/clippy.rs [hunk 2]
```diff
@@ -25,6 +28,17 @@ pub fn run_cmd(args: ClippyArgs) -> Result<()> {
...
 
+    // Pass along the `--message-format` flag if it was provided.
+    //
+    // We don't really care about validating that it's a valid argument to
+    // `cargo check --message-format`, because `cargo check` will error out if
+    // it's unrecognized, and repeating the validation here just presents an
+    // opportunity to get out of sync with what Cargo actually accepts should a
+    // new message format be added.
+    if let Some(fmt) = args.message_format {
+        command.args(&["--message-format", &fmt]);
```
Comment (sunshowers) [inline dev-tools/xtask/src/clippy.rs:39]: Don't need the first & here right?

**Review by sunshowers — APPROVED** [review]
Awesome! Just a couple small nits.

**Comment by hawkw** [comment]
I think @iliana's PR #5715 kind of obviates this (and honestly, kind of obviates `cargo xtask clippy` entirely, honestly), so I'll probably close this. But, if folks think this is worth having regardless, I'm happy to merge it instead.

**Comment by sunshowers** [comment]
I think I'd still want this tbh

**Comment by hawkw** [comment]
Mmkay, I'll get it ready to merge, then.

**Pull request merged**